homeassistant/components/fjaraskupan/__init__.py

```diff
@@ -0,0 +1,143 @@
+
+
+@dataclass
+class EntryState:
+    """Store state of config entry."""
+
+    scanner: BleakScanner
+    devices: dict[str, DeviceState]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Fjäråskupan from a config entry."""
+
+    scanner = BleakScanner()
+
+    state = EntryState(scanner, {})
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = state
+
+    async def detection_callback(
+        ble_device: BLEDevice, advertisement_data: AdvertisementData
+    ):
+        if device_filter(ble_device, advertisement_data):
+            _LOGGER.debug(
+                "Detection: %s %s - %s", ble_device.name, ble_device, advertisement_data
+            )
+
+            data = state.devices.get(ble_device.address)
+
+            if data:
+                data.device.detection_callback(ble_device, advertisement_data)
+                data.coordinator.async_set_updated_data(data.device.state)
+            else:
+
+                device = Device(ble_device)
+                device.detection_callback(ble_device, advertisement_data)
+
+                async def async_update_data():
+                    """Handle an explicit update request."""
+                    await device.update()
+                    return device.state
+
+                coordinator = DataUpdateCoordinator[State](
+                    hass,
+                    logger=_LOGGER,
+                    name="Fjäråskupan Updater",
+                    update_interval=timedelta(seconds=120),
+                    update_method=async_update_data,
+                )
+                coordinator.async_set_updated_data(device.state)
+
+                device_info: DeviceInfo = {
+                    "identifiers": {(DOMAIN, ble_device.address)},
+                    "manufacturer": "Fjäråskupan",
+                    "name": "Fjäråskupan",
+                }
+                data = DeviceState(device, coordinator, device_info)
+                state.devices[ble_device.address] = data
+                async_dispatcher_send(
+                    hass, DISPATCH_DETECTION, entry.entry_id, ble_device.address
+                )
+
+    scanner.register_detection_callback(detection_callback)
+    await scanner.start()
+
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+    return True
+
+
+async def async_setup_entry_platform(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+    constructor: Callable[[DeviceState], list[Entity]],
+) -> None:
+    """Set up a platform with added entities."""
+
+    entrystate: EntryState = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        [
+            entity
+            for devicestate in entrystate.devices.values()
+            for entity in constructor(devicestate)
+        ]
+    )
+
+    @callback
+    def _detection(entry_id: str, address: str):
+        if entry_id != entry.entry_id:
```

Comment:
If we include the entry id in the signal string we don't need this check.

Reply:
I've seen that done. It just feels like an odd way of doing it.

Reply:
Besides simplifying the callback here, it will optimize the calls in the dispatch helper.

Reply:
I've fixed this. I think we should extend dispatcher to keep a tuple instead of str as as key for the signals. That way we don't need to "construct" odd strings and get the same optimization for free.